File: ios/ReactNativeViewController.m
#import "ReactNativeViewController.h"
#import <React/RCTRootView.h>
#import <React/RCTBridge.h>
#import "ReactNativeBrownfield.h"
#import "ReactNativeBrownfieldNotifications.h"

@implementation ReactNativeViewController

@synthesize moduleName = _moduleName;
@synthesize initialProperties = _initialProperties;

-(instancetype)initWithModuleName:(NSString *)moduleName {
    return [self initWithModuleName:moduleName andInitialProperties:nil];
}

-(instancetype)initWithModuleName:(NSString *)moduleName
             andInitialProperties:(NSDictionary*)initialProperties {
    self = [super init];
    _moduleName = moduleName;
    _initialProperties = initialProperties;
    
    return self;
}

-(void)viewDidLoad {
    RCTRootViewFactory *factory = [[ReactNativeBrownfield shared] rootViewFactory];
    if (factory == nil) {
        NSLog(@"Error: You need to start React Native in order to use ReactNativeViewController, make sure to run [[BridgeManager shared] startReactNative] before instantiating it.");
        return;
    }

    if (_moduleName) {
        self.view = [factory viewWithModuleName:_moduleName initialProperties:_initialProperties];
        
        [[NSNotificationCenter defaultCenter]
            addObserver:self selector:@selector(togglePopGestureRecognizer:)
            name:TogglePopGestureRecognizerNotification object:nil];
        [[NSNotificationCenter defaultCenter]
         addObserver:self selector:@selector(popToNative:)
         name:PopToNativeNotification object:nil];
    }
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)togglePopGestureRecognizer:(NSNotification*)notification {
    NSDictionary *userInfo = notification.userInfo;
    BOOL enabled = [[userInfo objectForKey:@"enabled"] boolValue];
    dispatch_async(dispatch_get_main_queue(), ^{
        self.navigationController.interactivePopGestureRecognizer.enabled = enabled;
    });
}

- (void)popToNative:(NSNotification*)notification {
    NSDictionary *userInfo = notification.userInfo;
    BOOL animated = [[userInfo objectForKey:@"animated"] boolValue];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.navigationController popViewControllerAnimated:animated];
    });
}

@end

File: ios/ReactNativeViewController.swift
import UIKit
import React

@objc public class ReactNativeViewController: UIViewController {
  private var moduleName: String
  private var initialProperties: [String: Any]?
  
  @objc public init(moduleName: String, initialProperties: [String: Any]? = nil) {
    self.moduleName = moduleName
    self.initialProperties = initialProperties
    super.init(nibName: nil, bundle: nil)
  }
  
  required init?(coder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }
  
  public override func viewDidLoad() {
    super.viewDidLoad()
    
    guard let factory = ReactNativeBrownfield.shared.reactNativeFactory?.rootViewFactory else {
      print("Error: You need to start React Native in order to use ReactNativeViewController, make sure to run BridgeManager.shared.startReactNative() before instantiating it.")
      return
    }
    
    if !moduleName.isEmpty {
      view = factory.view(withModuleName: moduleName, initialProperties: initialProperties)
      
      NotificationCenter.default.addObserver(
        self,
        selector: #selector(togglePopGestureRecognizer(_:)),
        name: NSNotification.Name.togglePopGestureRecognizer,
        object: nil
      )
      
      NotificationCenter.default.addObserver(
        self,
        selector: #selector(popToNative(_:)),
        name: NSNotification.Name.popToNative,
        object: nil
      )
    }
  }
  
  deinit {
    NotificationCenter.default.removeObserver(self)
  }
  
  @objc private func togglePopGestureRecognizer(_ notification: Notification) {
    guard let userInfo = notification.userInfo,
          let enabled = userInfo["enabled"] as? Bool else { return }
    
    DispatchQueue.main.async { [weak self] in
      self?.navigationController?.interactivePopGestureRecognizer?.isEnabled = enabled
    }
  }
  
  @objc private func popToNative(_ notification: Notification) {
    guard let userInfo = notification.userInfo,
          let animated = userInfo["animated"] as? Bool else { return }
    
    DispatchQueue.main.async { [weak self] in
      self?.navigationController?.popViewController(animated: animated)
    }
  }
}
